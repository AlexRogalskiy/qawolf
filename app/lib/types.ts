// GraphQL Types
export type AuthenticatedUser = {
  access_token: string;
  user: User;
};

export type DeploymentEnvironment = "all" | "preview" | "production";

export type Env = { [name: string]: string };

export type Environment = {
  id: string;
  name: string;
};

export type EnvironmentVariable = {
  created_at: string;
  environment_id: string;
  id: string;
  name: string;
  value: string;
};

export type Integration = {
  github_repo_name: string;
  id: string;
  slack_channel: string | null;
  team_name: string | null;
  type: "github" | "slack";
};

export type Invite = {
  email: string;
  id: string;
  team_id: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  code: string;
  completed_at: string | null;
  created_at: string;
  current_line: number | null;
  environment_id: string | null;
  id: string;
  logs_url: string | null;
  started_at: string | null;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  video_url: string | null;
};

export type Runner = {
  api_key: string | null;
  ws_url: string | null;
};

export type ShortSuite = {
  created_at: string;
  environment_id: string | null;
  environment_variables: string | null;
  id: string;
  team_id: string;
  trigger_color: string | null;
  trigger_id: string | null;
  trigger_name: string | null;
};

export type Suite = ShortSuite & {
  runs: SuiteRun[];
};

export type SuiteRun = {
  completed_at: string | null;
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  started_at: string | null;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type ShortTeam = {
  id: string;
  name: string;
};

export type Team = ShortTeam & {
  alert_integration_id: string | null;
  api_key: string;
  helpers: string;
  inbox: string;
  is_email_alert_enabled: boolean;
  is_enabled: boolean;
  next_trigger_id: string;
  renewed_at: string | null;
};

export type ShortTest = {
  id: string;
  name: string;
};

export type Test = ShortTest & {
  code: string;
  deleted_at: string | null;
  is_enabled: boolean;
  updated_at: string;
  version: number;
};

export type TestHistoryRun = {
  created_at: string;
  id: string;
  started_at: string;
  status: RunStatus;
};

export type TestSummaryRun = {
  created_at: string;
  gif_url: string | null;
  id: string;
  status: RunStatus;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: TestSummaryRun[];
  test_id: string;
};

export type Trigger = {
  color: string;
  created_at: string;
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  id: string;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
};

export type User = {
  avatar_url: string | null;
  email: string;
  id: string;
  onboarded_at: string | null;
  teams: ShortTeam[];
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Component Types
export type AuthMode = "logIn" | "signUp";

export type CreateCode = {
  code: string;
  test_id: string | null;
};

export type Modal =
  | "createTest"
  | "environments"
  | "deleteTests"
  | "teamSettings"
  | "triggers";

export type NavigationOption = "code" | "logs" | "helpers";

export type NavigationType = "dark" | "light";

export type SelectedTest = {
  id: string;
  name: string;
};

export type Side = "left" | "right";

export type TestTriggers = {
  test_id: string;
  trigger_ids: string[];
};

export type TriggerFields = {
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  name: string;
  repeat_minutes: number | null;
};

export type ValueProp = {
  detail: string;
  message: string;
};

export type Wolf = {
  name: string;
  number: number;
  variant: string;
};

// Runner Types
export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id: string;
  version: number;
};

export type Log = {
  message: string;
  severity: "info" | "warning" | "error";
  timestamp: string;
};

export type RunOptions = {
  code: string;
  end_line?: number;
  env: Env;
  helpers?: string;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id?: string;
  version: number;
};

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
};

// State Types
type SignUp = {
  inviteId?: string | null;
  redirectUri?: string | null;
};

type ModalState = {
  name: Modal | null;
  teamId?: string;
  testIds?: string[];
  tests?: SelectedTest[];
};

export type State = {
  dashboardUri: string | null;
  editorSidebarWidth: number;
  email: string | null;
  environmentId: string | null;
  modal: ModalState;
  run: RunOptions | null;
  signUp: SignUp;
  teamId: string | null;
  toast: {
    error?: boolean;
    expiresIn?: number;
    message: string;
  } | null;
};
