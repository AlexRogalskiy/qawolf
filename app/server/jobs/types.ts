export const JOB_TYPES = [
  "checkPending",
  "deleteOldEmails",
  "deleteRunners",
  "deployRunners",
  "orchestrateRunners",
  "orchestrateTriggers",
  "restartRunners",
] as const;

export type Job = typeof JOB_TYPES[number];
