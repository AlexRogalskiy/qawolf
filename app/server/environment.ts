/* eslint-disable @typescript-eslint/no-non-null-assertion */
import dotenv from "dotenv";

dotenv.config(process.env.DOTENV_PATH ? { path: process.env.DOTENV_PATH } : {});

const databaseSsl = process.env.DATABASE_SSL;

export default {
  APP_URL: process.env.APP_URL!,
  DATABASE_SSL: databaseSsl ? JSON.parse(databaseSsl) : null,
  DATABASE_URL: process.env.DATABASE_URL,
  EMAIL_DOMAIN: process.env.EMAIL_DOMAIN!,
  ENCRYPT_IV: process.env.ENCRYPT_IV,
  ENCRYPT_KEY: process.env.ENCRYPT_KEY,
  GITHUB_APP_ID: process.env.GITHUB_APP_ID!,
  GITHUB_OAUTH_CLIENT_ID: process.env.GITHUB_OAUTH_CLIENT_ID!,
  GITHUB_OAUTH_CLIENT_SECRET: process.env.GITHUB_OAUTH_CLIENT_SECRET!,
  GITHUB_SYNC_APP_ID: process.env.GITHUB_SYNC_APP_ID!,
  GITHUB_USER_AGENT: process.env.GITHUB_USER_AGENT!,
  GITHUB_WEBHOOK_SECRET: process.env.GITHUB_WEBHOOK_SECRET!,
  INTERCOM_IDENTITY_SECRET: process.env.INTERCOM_IDENTITY_SECRET || "secret",
  IPSTACK_API_KEY: process.env.IPSTACK_API_KEY!,
  JOB_SECRET: process.env.JOB_SECRET!,
  JWT_EXPIRATION: process.env.JWT_EXPIRATION!,
  JWT_SECRET: process.env.JWT_SECRET!,
  NODE_ENV: process.env.NODE_ENV || "development",
  QAWOLF_AWS_ACCESS_KEY_ID: process.env.QAWOLF_AWS_ACCESS_KEY_ID!,
  QAWOLF_AWS_REGION: process.env.QAWOLF_AWS_REGION!,
  QAWOLF_AWS_SECRET_ACCESS_KEY: process.env.QAWOLF_AWS_SECRET_ACCESS_KEY!,
  QAWOLF_AWS_S3_BUCKET: process.env.QAWOLF_AWS_S3_BUCKET!,
  SEGMENT_WRITE_KEY: process.env.SEGMENT_WRITE_KEY || "",
  SENDGRID_API_KEY: process.env.SENDGRID_API_KEY!,
  SENDGRID_WEBHOOK_SECRET: process.env.SENDGRID_WEBHOOK_SECRET!,
  SLACK_CLIENT_ID: process.env.SLACK_CLIENT_ID!,
  SLACK_CLIENT_SECRET: process.env.SLACK_CLIENT_SECRET!,
  SLACK_UPDATES_WEBHOOK: process.env.SLACK_UPDATES_WEBHOOK || null,
  STRIPE_API_KEY: process.env.STRIPE_API_KEY!,
  STRIPE_BASE_PRICE_ID: process.env.STRIPE_BASE_PRICE_ID!,
  STRIPE_METERED_PRICE_ID: process.env.STRIPE_METERED_PRICE_ID!,
  STRIPE_WEBHOOK_SECRET: process.env.STRIPE_WEBHOOK_SECRET!,
  VERCEL_GIT_COMMIT_SHA: process.env.VERCEL_GIT_COMMIT_SHA || "",
  VERCEL_URL: process.env.VERCEL_URL
    ? `https://${process.env.VERCEL_URL}`
    : null,
};
