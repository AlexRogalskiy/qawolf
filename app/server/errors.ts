import {
  ApolloError,
  AuthenticationError as ApolloAuthenticationError,
} from "apollo-server-micro";

// These errors are forwarded to the client / user
export const CLIENT_ERROR_CODES = ["CLIENT", "UNAUTHENTICATED"];

export class AuthenticationError extends ApolloAuthenticationError {
  constructor(message = "unauthorized") {
    super(message);
  }
}

export class ClientError extends ApolloError {
  constructor(message: string) {
    super(message, "CLIENT");

    Object.defineProperty(this, "name", { value: "ClientError" });
  }
}
