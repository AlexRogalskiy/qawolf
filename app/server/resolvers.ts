import {
  apiKeysResolver,
  createApiKeyResolver,
  deleteApiKeyResolver,
} from "./resolvers/api_key";
import { createGitHubSignInUrlResolver } from "./resolvers/auth";
import { dashboardResolver } from "./resolvers/dashboard";
import {
  createEnvironmentResolver,
  deleteEnvironmentResolver,
  environmentsResolver,
  updateEnvironmentResolver,
} from "./resolvers/environment";
import {
  createEnvironmentVariableResolver,
  deleteEnvironmentVariableResolver,
  environmentVariablesResolver,
  updateEnvironmentVariableResolver,
} from "./resolvers/environment_variable";
import { createGitHubIntegrationsResolver } from "./resolvers/github";
import {
  createTriggerResolver,
  deleteTriggerResolver,
  testTriggersResolver,
  triggersResolver,
  updateTriggerResolver,
} from "./resolvers/trigger";
import { updateTestTriggersResolver } from "./resolvers/test_trigger";
import { integrationsResolver } from "./resolvers/integration";
import {
  acceptInviteResolver,
  createInvitesResolver,
  teamInvitesResolver,
} from "./resolvers/invite";
import {
  suiteRunsResolver,
  testHistoryResolver,
  updateRunResolver,
} from "./resolvers/run";
import { runnerResolver, updateRunnerResolver } from "./resolvers/runner";
import { joinMailingListResolver } from "./resolvers/sendgrid";
import {
  createSlackIntegrationResolver,
  createSlackIntegrationUrlResolver,
} from "./resolvers/slack";
import {
  createSuiteResolver,
  suiteResolver,
  suitesResolver,
} from "./resolvers/suite";
import { teamResolver, updateTeamResolver } from "./resolvers/team";
import {
  createTestResolver,
  deleteTestsResolver,
  testResolver,
  testsResolver,
  testSummaryResolver,
  updateTestResolver,
} from "./resolvers/test";
import {
  currentUserResolver,
  sendLoginCodeResolver,
  signInWithEmailResolver,
  signInWithGitHubResolver,
  teamUsersResolver,
  updateUserResolver,
} from "./resolvers/user";

export const resolvers = {
  Dashboard: {
    suites: suitesResolver,
    tests: testsResolver,
  },
  Team: {
    invites: teamInvitesResolver,
    users: teamUsersResolver,
  },
  Test: {
    summary: testSummaryResolver,
    triggers: testTriggersResolver,
  },
  Suite: {
    runs: suiteRunsResolver,
  },
  Mutation: {
    acceptInvite: acceptInviteResolver,
    createApiKey: createApiKeyResolver,
    createEnvironment: createEnvironmentResolver,
    createEnvironmentVariable: createEnvironmentVariableResolver,
    createGitHubIntegrations: createGitHubIntegrationsResolver,
    createInvites: createInvitesResolver,
    createSignInUrl: createGitHubSignInUrlResolver,
    createSlackIntegration: createSlackIntegrationResolver,
    createSlackIntegrationUrl: createSlackIntegrationUrlResolver,
    createSuite: createSuiteResolver,
    createTest: createTestResolver,
    createTrigger: createTriggerResolver,
    deleteApiKey: deleteApiKeyResolver,
    deleteEnvironment: deleteEnvironmentResolver,
    deleteEnvironmentVariable: deleteEnvironmentVariableResolver,
    deleteTests: deleteTestsResolver,
    deleteTrigger: deleteTriggerResolver,
    joinMailingList: joinMailingListResolver,
    sendLoginCode: sendLoginCodeResolver,
    signInWithEmail: signInWithEmailResolver,
    signInWithGitHub: signInWithGitHubResolver,
    updateEnvironment: updateEnvironmentResolver,
    updateEnvironmentVariable: updateEnvironmentVariableResolver,
    updateRun: updateRunResolver,
    updateRunner: updateRunnerResolver,
    updateTeam: updateTeamResolver,
    updateTest: updateTestResolver,
    updateTestTriggers: updateTestTriggersResolver,
    updateTrigger: updateTriggerResolver,
    updateUser: updateUserResolver,
  },
  Query: {
    apiKeys: apiKeysResolver,
    currentUser: currentUserResolver,
    dashboard: dashboardResolver,
    environments: environmentsResolver,
    environmentVariables: environmentVariablesResolver,
    integrations: integrationsResolver,
    runner: runnerResolver,
    suite: suiteResolver,
    team: teamResolver,
    test: testResolver,
    testHistory: testHistoryResolver,
    triggers: triggersResolver,
  },
};
