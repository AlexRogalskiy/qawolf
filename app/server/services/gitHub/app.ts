import { createAppAuth } from "@octokit/auth-app";
import { Octokit } from "@octokit/rest";

import environment from "../../environment";
import { findSystemEnvironmentVariable } from "../../models/environment_variable";
import { findIntegration } from "../../models/integration";
import { ModelOptions } from "../../types";

type InstallationOptions = {
  installationId: number;
  isSync?: boolean;
};

type OctokitResult = {
  octokit: Octokit;
  token: string;
};

export type OctokitRepo = {
  octokit: Octokit;
  owner: string;
  repo: string;
};

export const createInstallationAccessToken = async (
  { installationId, isSync }: InstallationOptions,
  options: ModelOptions
): Promise<string> => {
  const privateKeyVariable = await findSystemEnvironmentVariable(
    isSync ? "GITHUB_SYNC_APP_PRIVATE_KEY" : "GITHUB_APP_PRIVATE_KEY",
    options
  );
  const clientSecretVariable = await findSystemEnvironmentVariable(
    isSync ? "GITHUB_SYNC_APP_CLIENT_SECRET" : "GITHUB_APP_CLIENT_SECRET",
    options
  );

  const appId = isSync
    ? environment.GITHUB_SYNC_APP_ID
    : environment.GITHUB_APP_ID;

  const auth = createAppAuth({
    appId,
    clientId: environment.GITHUB_OAUTH_CLIENT_ID,
    clientSecret: clientSecretVariable.value,
    installationId,
    privateKey: JSON.parse(privateKeyVariable.value),
  });

  const { token } = await auth({ type: "installation" });
  return token;
};

export const createOctokitForInstallation = async (
  { installationId, isSync }: InstallationOptions,
  options: ModelOptions
): Promise<OctokitResult> => {
  const token = await createInstallationAccessToken(
    { installationId, isSync },
    options
  );
  const octokit = new Octokit({ auth: token });
  return { octokit, token };
};

export const createOctokitForIntegration = async (
  integrationId: string,
  options: ModelOptions
): Promise<OctokitRepo> => {
  const integration = await findIntegration(integrationId, options);

  const result = await createOctokitForInstallation(
    {
      installationId: integration.github_installation_id,
      isSync: integration.type === "github_sync",
    },
    options
  );

  const [owner, repo] = integration.github_repo_name?.split("/");
  return { ...result, owner, repo };
};
