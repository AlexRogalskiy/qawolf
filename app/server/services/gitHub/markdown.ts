import environment from "../../environment";
import { SuiteRun, Trigger, User } from "../../types";

type BuildCommentForSuite = {
  runs: SuiteRun[];
  suite_id: string;
  trigger: Trigger;
  user: User;
};

const buildDetailsForRuns = (runs: SuiteRun[]): string => {
  if (!runs.length) return "";

  const status = runs[0].status;
  const emoji = status === "fail" ? "‚ùå" : "‚úÖ";

  const runsHtml = runs
    .map((run) => {
      const runUrl = new URL(`/run/${run.id}`, environment.APP_URL).href;

      return `<tr>
<td>
<div>
<a href="${runUrl}"><p>${emoji} ${run.test_name}</p></a>
<img src="${run.gif_url}" width="240" />
</div>
</td>
</tr>`;
    })
    .join("\n");

  return `
<details>
<summary>${emoji} ${runs.length} ${status}</summary>
<table>
${runsHtml}
</table>
</details>`;
};

export const buildCommentForSuite = ({
  runs,
  suite_id,
  trigger,
  user,
}: BuildCommentForSuite): string => {
  const suiteUrl = new URL(`/suites/${suite_id}`, environment.APP_URL).href;
  let emoji = "üê∫";

  const inProgressRuns = runs.filter((r) => r.status === "created");
  const failingRuns = runs.filter((r) => r.status === "fail");
  const passingRuns = runs.filter((r) => r.status === "pass");

  if (!inProgressRuns.length) emoji = failingRuns.length ? "‚ùå" : "‚úÖ";
  const isPass = passingRuns.length === runs.length;

  const header = inProgressRuns.length
    ? `${inProgressRuns.length} test${
        inProgressRuns.length === 1 ? "" : "s"
      } running`
    : `${runs.length} test${runs.length === 1 ? "" : "s"} ran`;

  return `## ${emoji} QA Wolf - ${trigger.name}
<img src="https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-${
    user.wolf_variant
  }${isPass ? "-party" : ""}-slack.png" width="32" /> ${
    user.wolf_name
  } here: ${header}, [see details here](${suiteUrl}).${buildDetailsForRuns(
    failingRuns
  )}${buildDetailsForRuns(passingRuns)}`;
};
