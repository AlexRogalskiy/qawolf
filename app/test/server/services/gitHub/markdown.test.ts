import { buildCommentForSuite } from "../../../../server/services/gitHub/markdown";
import { SuiteRun, Trigger, User } from "../../../../server/types";

const failingRun = {
  id: "runId",
  gif_url: "https://media.giphy.com/media/10LKovKon8DENq/giphy.gif",
  status: "fail",
  test_name: "Subscribe",
};

const inProgressRun = {
  id: "run2Id",
  gif_url: "https://media.giphy.com/media/10LKovKon8DENq/giphy.gif",
  status: "created",
  test_name: "Log in",
};

const passingRun = {
  id: "run3Id",
  gif_url: "https://media.giphy.com/media/10LKovKon8DENq/giphy.gif",
  status: "pass",
  test_name: "Sign up",
};

describe("buildCommentForSuite", () => {
  it("builds a comment for single running test", () => {
    expect(
      buildCommentForSuite({
        runs: [inProgressRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Wolverina", wolf_variant: "black" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for multiple running tests", () => {
    expect(
      buildCommentForSuite({
        runs: [inProgressRun, inProgressRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for single passing test", () => {
    expect(
      buildCommentForSuite({
        runs: [passingRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for multiple passing tests", () => {
    expect(
      buildCommentForSuite({
        runs: [passingRun, passingRun, passingRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for single failing test", () => {
    expect(
      buildCommentForSuite({
        runs: [failingRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for multiple failing tests", () => {
    expect(
      buildCommentForSuite({
        runs: [failingRun, failingRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });

  it("builds a comment for suites with tests of all statuses", () => {
    expect(
      buildCommentForSuite({
        runs: [failingRun, passingRun, failingRun, inProgressRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();

    expect(
      buildCommentForSuite({
        runs: [failingRun, failingRun, inProgressRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();

    expect(
      buildCommentForSuite({
        runs: [passingRun, inProgressRun, passingRun, passingRun] as SuiteRun[],
        suite_id: "suiteId",
        trigger: { name: "Deployment" } as Trigger,
        user: { wolf_name: "Bean", wolf_variant: "husky" } as User,
      })
    ).toMatchSnapshot();
  });
});
