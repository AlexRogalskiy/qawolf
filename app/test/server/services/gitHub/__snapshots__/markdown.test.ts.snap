// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCommentForSuite builds a comment for multiple failing tests 1`] = `
"## ❌ QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 2 tests ran, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>❌ 2 fail</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for multiple passing tests 1`] = `
"## ✅ QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 3 tests ran, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>✅ 3 pass</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for multiple running tests 1`] = `
"## 🐺 QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 2 tests running, [see details here](http://localhost:3000/suites/suiteId)."
`;

exports[`buildCommentForSuite builds a comment for single failing test 1`] = `
"## ❌ QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 1 test ran, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>❌ 1 fail</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for single passing test 1`] = `
"## ✅ QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 1 test ran, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>✅ 1 pass</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for single running test 1`] = `
"## 🐺 QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-black-slack.png\\" width=\\"24\\" /> Wolverina here: 1 test running, [see details here](http://localhost:3000/suites/suiteId)."
`;

exports[`buildCommentForSuite builds a comment for suites with tests of all statuses 1`] = `
"## 🐺 QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 1 test running, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>❌ 2 fail</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>
<details>
<summary>✅ 1 pass</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for suites with tests of all statuses 2`] = `
"## 🐺 QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 1 test running, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>❌ 2 fail</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/runId\\"><p>❌ Subscribe</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;

exports[`buildCommentForSuite builds a comment for suites with tests of all statuses 3`] = `
"## 🐺 QA Wolf - Deployment
<img src=\\"https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-husky-slack.png\\" width=\\"24\\" /> Bean here: 1 test running, [see details here](http://localhost:3000/suites/suiteId).
<details>
<summary>✅ 3 pass</summary>
<table>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
<tr>
<td>
<div>
<a href=\\"http://localhost:3000/run/run3Id\\"><p>✅ Sign up</p></a>
<img src=\\"https://media.giphy.com/media/10LKovKon8DENq/giphy.gif\\" width=\\"240\\" />
</div>
</td>
</tr>
</table>
</details>"
`;
